simple_certmanager/forms.py

```diff
@@ -13,6 +15,43 @@
 from .validators import PrivateKeyValidator
 
 
+class SigningRequestAdminForm(forms.ModelForm):
+    should_renew_csr = forms.BooleanField(
+        label=_("Regenerate CSR"),
+        help_text=_(
+            "Check this box to regenerate the CSR from the above details. "
+            "This will also generate a new private key."
+        ),
+        required=False,
+    )
+
+    class Meta:
+        model = SigningRequest
+        fields = "__all__"
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # Show the "Regenerate CSR" checkbox when editing an existing SigningRequest
+        if not self.instance.pk:
+            self.fields.pop("should_renew_csr")
+
+    def clean_should_renew_csr(self):
+        # Regenerate the CSR/Private Key if the checkbox is checked
```

Comment:
you really have to see validation and processing as two separate things - first everything must be validated before you can process based on the input data.

This implementation would persist changes while other fields may be invalid, so then you get validation errors but also changes will be applied and that's mega confusing.

During validation, you may *only* validate. Do not touch `self.instance`, and especially don't write to the database or filesystem. Instead, validate everything, and once validation passes, you process the changes in `def save`.